### DISCUSSION TITLE
`./run` spends a lot of time compiling CLI after switching branch

### DISCUSSION THREAD

**Discussion by somebody1234** [enso-org/enso#6230] [Help and troubleshooting] [answered]
For me (I'm using Linux), every time I switch branch, the following four binaries are rebuilt, every single time:
```
enso_build_cli
enso_remove_draft_releases
enso_disable_wasm_opt
enso_build4
```
The process takes up to 30 minutes - I'm not sure whether this is a Linux-only issue, or whether I've set up my environment wrong.
I'm using the normal command to build, if that helps:
```
./run ide build --skip-version-check --skip-wasm-opt --backend-source release --backend-release latest
```
(Note also the sizes - are they all supposed to be ~500MB?)
```lua
-rwxr-xr-x 2 me me  582M Mar 23 23:23 enso_build_cli-d493a23d94f2e700
-rwxr-xr-x 2 me me  560M Mar 23 23:23 enso_remove_draft_releases-5f6afebbccda6fa2
-rwxr-xr-x 2 me me  551M Mar 23 23:23 enso_disable_wasm_opt-1e087024e9a32d1b
-rwxr-xr-x 2 me me  455M Mar 23 23:23 enso_build4-9b4fafa26c07d4e4
```

**Comment by somebody1234** [comment]
@mwu-tow I was told you would be the person to contact about this - do you know why this is happening (or whether this is supposed to happen)?

**Reply by mwu-tow** [reply]
@somebody1234 

So there are several questions.

1. Why all four binaries are built. Actually, it should be enough to build `enso_build_cli` for our usecases, adding appropriate flags to our `run.*`  scripts would help this.
2. Building the build script itself should not take 30 minutes.
  * What is your hardware?
  * First of all, building build script should be significantly faster than building GUI itself. Would you be able to measure how much time it takes to build the build script itself (`cargo build --profile buildscript --target-dir "target/enso-build" --package enso-build-cli`) and then how much time it takes to build IDE (`./run ide build`) on a clean working copy?
 3. The binaries are heavy likely due to the debug symbols (which are really handy when you need to debug the build script itself). While this is a bit wasteful (particularly since additional executables are present, as per `1.`), it still should be dominated by the size of GUI binaries.

**Comment by mwu-tow** [comment]
@somebody1234 
> For me (I'm using Linux), every time I switch branch, the following four binaries are rebuilt, every single time:

The build script should not recompile, unless there is a change in sources. I'd expect that it was either modified, or some of its dependencies were modified (as it shared the code with some parts of GUI). But if there was no change in dependencies, the build script should not recompile.

Some team members `cargo install` the build script and then invoke it manually through `enso-build-cli`. While it is not "officially supported" it should work as long as there are no breaking changes in the build logic.

**Reply by somebody1234** [reply]
re: should not recompile - i don't know. it happens for me (literally) every time i switch branches though.
but yesterday i looked at CONTRIBUTING.md and realized that i wasn't using cargo-watch-plus, so i switched to that yesterday.
maybe that was the cause of the problem? (say, incorrectly reading gitignores?)

**Reply by somebody1234** [reply]
yeah, after switching branches *after* changing to cargo-watch-plus, i can no longer repro this issue.

**Reply by somebody1234** [reply]
(it no longer rebuilds - seems like worst case is now ~16s, regular case is now ~0.3s)

**Comment by somebody1234** [comment, marked as answer]
So it turns out it was an issue with `cargo watch` - maybe the importance of using `cargo-watch-plus` should be highlighted in CONTRIBUTING.md?

I think I more or less did the bare minimum to set things up, because I wasn't expecting to have to compile Rust sources

**Reply by wdanilo** [reply]
In fact, we should switch to cargo-watch-plus by default. It looks like it works for everyone else, we might give it a shot, what do you think guys about it? CC @mwu-tow @farmaazon
